File: apps/ios/ScribbleFit/Sources/Features/Canvas/Domain/UseCases/ConfirmScribbleUseCase.swift
import Foundation
#if SWIFT_PACKAGE
import CoreModel
#endif

@MainActor
public final class ConfirmScribbleUseCase {
    private let scribbleRepository: ScribbleRepository
    private let workoutRepository: WorkoutRepository

    public init(scribbleRepository: ScribbleRepository, workoutRepository: WorkoutRepository) {
        self.scribbleRepository = scribbleRepository
        self.workoutRepository = workoutRepository
    }

    public func execute(scribble: Scribble) async throws {
        guard scribble.status == .success else {
            throw ConfirmScribbleError.invalidStatus(scribble.status)
        }

        // Deep-copy exercises with fresh UUIDs to avoid SwiftData unique
        // constraint conflicts with the scribble's exercise entities.
        let freshExercises = scribble.exercises.map { exercise in
            Exercise(
                id: UUID(),
                canonicalName: exercise.canonicalName,
                muscleGroup: exercise.muscleGroup,
                sets: exercise.sets.map { set in
                    ExerciseSet(
                        id: UUID(),
                        setNumber: set.setNumber,
                        weight: set.weight,
                        reps: set.reps,
                        rpe: set.rpe,
                        notes: set.notes
                    )
                },
                isDraft: false,
                estimated1RM: exercise.estimated1RM,
                intensity: exercise.intensity,
                improvement: exercise.improvement
            )
        }

        let workout = Workout(
            id: UUID(),
            date: scribble.createdAt,
            exercises: freshExercises,
            notes: ["Imported from scribble: \(scribble.rawText)"]
        )

        try await workoutRepository.saveWorkout(workout)

        var completedScribble = scribble
        completedScribble.status = .completed
        try await scribbleRepository.updateScribble(completedScribble)
    }
}

public enum ConfirmScribbleError: Error {
    case invalidStatus(ScribbleStatus)
}

File: apps/android/feature/scribble/domain/src/main/java/com/scribblefit/feature/scribble/domain/usecase/ConfirmScribbleUseCase.kt
package com.scribblefit.feature.scribble.domain.usecase

import com.scribblefit.core.common.runCatchingWithCancellation
import com.scribblefit.core.model.Scribble
import com.scribblefit.core.model.ScribbleStatus
import com.scribblefit.feature.scribble.domain.ScribbleRepository
import com.scribblefit.feature.scribble.domain.error.ScribbleError
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.withContext

class ConfirmScribbleUseCase(
    private val scribbleRepository: ScribbleRepository,
    private val coroutineDispatcher: CoroutineDispatcher
) {
    suspend operator fun invoke(scribble: Scribble) = withContext(coroutineDispatcher) {
        runCatchingWithCancellation {
            if (scribble.status != ScribbleStatus.SUCCESS) {
                throw ScribbleError.InvalidStatus(scribble.status)
            }

            scribbleRepository.updateScribble(scribble.copy(status = ScribbleStatus.COMPLETED))
        }
    }
}
